### PR TITLE
Typer cli

### PR DESCRIPTION
**Pull request by nainakhan123**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1 @@
 .env
-
-.vscode
```
Comment (salmanmaq) [inline .gitignore:3]: I think we can keep `.vscode` in gitignore

File: train.py [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
-
-    train_yolo(args.weights, args.data_config, args.epochs)
+@app.command()
+def train_yolo(
+    model_path: str = typer.Argument(),
+    data_path: str = typer.Argument(),
+    epochs: int = typer.Option(),
+):
+    print("start")
+    model = YOLO(model_path)
```
Comment (salmanmaq) [inline train.py:15]: Let's add type hints for all instances we are initializing a variable

File: test.py [hunk 3]
```diff
@@ -1,22 +1,20 @@
...
-    parser.add_argument("path_to_test")
+app = typer.Typer()
 
-    args = parser.parse_args()
 
-    predict_yolo(args.path_to_weights, args.path_to_test)
+@app.command()
+def test_yolo(
+    weights_path: str = typer.Argument(),
+    test_path: str = typer.Argument(),
```
Comment (salmanmaq) [inline test.py:11]: Name of variable `test_path` is unclear. Let's make the name more specific and probably also add a description to these arguments if possible

File: train.py [hunk 4]
```diff
@@ -1,25 +1,23 @@
...
-    parser.add_argument("data-config", required=True)
-    parser.add_argument("epochs", type=int, default=2)
 
-    args = parser.parse_args()
-
-    train_yolo(args.weights, args.data_config, args.epochs)
+@app.command()
+def train_yolo(
+    model_path: str = typer.Argument(),
+    data_path: str = typer.Argument(),
```
Comment (salmanmaq) [inline train.py:11]: Name of variable can be more explicit

**Review by salmanmaq — COMMENTED** [review]
2 more things other than the specific comments:

1. Let's make a new few called `cli.py` and that should have all the logic for the command line interface. It can use functions from `train.py` and `test.py`. The cli could take an argument such as `--train` and `--test` and then run the respective logic
2. Let's add usage instructions in `README.md`

**Comment by nainakhan123** [comment]
I have made the changes please review it. Let me know if anything else is going to change.

File: cli.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+from train import train_yolo
+
+
+app = typer.Typer()
+
+
+@app.command()
+def train(
+    pre_trained_model_path: str = typer.Argument(),
+    data_to_be_trained_path: str = typer.Argument(),
```
Comment (salmanmaq) [inline cli.py:12]: This is a bit unclear whether this is a folder path or path to a config file

File: README.md [hunk 6]
```diff
@@ -3,17 +3,17 @@ Barcode-detection
...
 Description
-Here are the 5 steps for this project :
+Here are the 4 steps for this project :
 
 1. Implement YOLO version 8 from ultralystics, that is used for object detection.
 2. Next create an instance of yolo and load pre-trained model of yolov8.
-3. Call the train method of the YOLO model instance. This initiates the training process using a custom dataset defined in the provided YAML file.
-4. After training call the validate method of the model.
-5. When the model is trained, run te test.py to test the model on test data.
+3. Now through CLI run the command "python train.py path_to_trained_model path_to_data_to_train --epochs 2", replace the placeholders path_to_trained_model and path_to_data_to_train with your actual paths and train the model.
```
Comment (salmanmaq) [inline README.md:10]: For code we can use these characters \` for inline code blocks
`python train.py ....`

Or also ```

```
This a new line code block
```

File: .gitignore (hunk 1 above)
Comment (salmanmaq) [reply .gitignore:3]: Need to keep .vscode in gitingore

**Review by salmanmaq — COMMENTED** [review]
You can address the small changes I mentioned. Then we can test it together and hopefully merge it by today.

File: README.md [hunk 7]
```diff
@@ -3,17 +3,16 @@ Barcode-detection
...
 Description
-Here are the 5 steps for this project :
+Here are the 4 steps for this project :
 
 1. Implement YOLO version 8 from ultralystics, that is used for object detection.
 2. Next create an instance of yolo and load pre-trained model of yolov8.
-3. Call the train method of the YOLO model instance. This initiates the training process using a custom dataset defined in the provided YAML file.
-4. After training call the validate method of the model.
-5. When the model is trained, run te test.py to test the model on test data.
+3. Now through CLI run the command ```bash python cli.py --pre_trained_model_path model.pt --dataset_to_be_trained_config_path  --epochs 2```, replace the placeholders path_to_trained_model and path_to_data_to_train with your actual paths and train the model.
```
Comment (salmanmaq) [inline README.md:10]: You can move the run commands to new line

File: README.md [hunk 8]
```diff
@@ -3,17 +3,16 @@ Barcode-detection
...
-Here are the 5 steps for this project :
+Here are the 4 steps for this project :
 
 1. Implement YOLO version 8 from ultralystics, that is used for object detection.
 2. Next create an instance of yolo and load pre-trained model of yolov8.
-3. Call the train method of the YOLO model instance. This initiates the training process using a custom dataset defined in the provided YAML file.
-4. After training call the validate method of the model.
-5. When the model is trained, run te test.py to test the model on test data.
+3. Now through CLI run the command ```bash python cli.py --pre_trained_model_path model.pt --dataset_to_be_trained_config_path  --epochs 2```, replace the placeholders path_to_trained_model and path_to_data_to_train with your actual paths and train the model. 
+4. After training the model, test the performance on test dataset running the the command through CLI ```python cli.py --path_to_best weights weights.pt --path_to_test_data test_data```. Replace the placeholder with your dataset path.
```
Comment (salmanmaq) [inline README.md:11]: Move run command to new line

File: cli.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+app = typer.Typer()
+
+
+@app.command()
+def train(
+    pre_trained_model_path: str = typer.Option(
+        "--path_to_pretrained_model", help="Path to pre-trained model"
+    ),
+    data_to_be_trained_config_path: str = typer.Option(
+        "--data_to_be_trained_config_path", help="Path to dataset to be train"
```
Comment (salmanmaq) [inline cli.py:16]: We can rename it to `dataset_config_path`. Also update the `help` to reflect that

File: train.py [hunk 10]
```diff
@@ -1,25 +1,14 @@
...
 
 
-def train_yolo(weights_path, data_config, epochs):
-    print("Starting training...")
-    model = YOLO(weights_path)
-    model.train(data=data_config, epochs=epochs)
-    print("Training completed.")
+def train_yolo(
+    pre_trained_model_path: str,
+    dataset_to_be_trained_config_path: str,
```
Comment (salmanmaq) [inline train.py:6]: Rename to `dataset_config_path`

**Pull request merged**